BingePlay Streaming Analytics Project

Name: Tanmay Purushottam Bokade<br>
Roll Number: TANMAY8604123<br>
Batch: July 2026<br>
Date: 17 Aug 2026<br>

In [ ]:
import pandas as pd
from sqlalchemy import create_engine

db_url = "mysql+pymysql://root:tanmay@localhost:3306/bingeplay"
engine = create_engine(db_url)
print("Connected to BingePlay database successfully!")

In [ ]:
# Verifying the NULL trap
verify_query = """
SELECT COUNT(*) AS null_user_count
FROM watch_sessions
WHERE user_id IS NULL;
"""
df_null_check = pd.read_sql(verify_query, engine)
print("--- NULL Trap Verification ---")
print(df_null_check)

--- NULL Trap Verification ---
   null_user_count
0                2


In [ ]:
query_q1 = """
SELECT
    COUNT(subscription_id) AS active_subs,
    SUM(monthly_price_inr) AS total_monthly_rev
FROM subscriptions
WHERE status = 'active'
  AND (end_date IS NULL OR end_date > '2024-06-30');
"""

df_q1 = pd.read_sql(query_q1, engine)

print("--- Q1: Active Revenue ---")
print(df_q1)

--- Q1: Active Revenue ---
   active_subs  total_monthly_rev
0         2340           784260.0


#### Q1 — Revenue of active subs<br>
**Answer: 2340 active subscriptions**<br>
**Total monthly revenue: ₹ 784,260**

Counted active subs.
Summed monthly prices.
Checked end_date.
NULL or future date means still active.


In [ ]:
query_q2 = """
SELECT
    MONTH(signup_date) AS month,
    MONTHNAME(signup_date) AS month_name,
    COUNT(user_id) AS signup_count
FROM users
WHERE YEAR(signup_date) = 2024
GROUP BY month, month_name
ORDER BY month;
"""

df_q2 = pd.read_sql(query_q2, engine)

print("--- Q2: Signup Momentum ---")
print(df_q2)

# Find the month with the highest signups
top_month = df_q2.loc[df_q2['signup_count'].idxmax(), 'month_name']
print(f"\nMonth with highest signups: {top_month}")

# Find ALL months with the highest signups (Assisted ai)
max_signups = df_q2['signup_count'].max()
top_months = df_q2[df_q2['signup_count'] == max_signups]['month_name'].tolist()
print(f"\nMonth(s) with highest signups: {', '.join(top_months)}")

--- Q2: Signup Momentum ---
   month month_name  signup_count
0      1    January           350
1      2   February           400
2      3      March           500
3      4      April           550
4      5        May           600
5      6       June           600

Month with highest signups: May

Month(s) with highest signups: May, June


#### Q2 — Signup momentum
**Answer:**
* January: 350
* February: 400
* March: 500
* April: 550
* May: 600
* June: 600

**Highest signups month:** May and June (Tie with 600)

Extracted month from signup_date.
Grouped users by month.
Counted total signups per group.
Sorted chronologically to track momentum.
Accounted for ties in the maximum count.

In [ ]:
query_q3 = """
SELECT
    device_type,
    COUNT(session_id) AS total_sessions,
    SUM(watch_minutes) AS total_watch_min,
    ROUND(AVG(watch_minutes), 2) AS avg_watch_min,
    ROUND(AVG(completed) * 100, 2) AS completion_rate
FROM watch_sessions
WHERE user_id IS NOT NULL
GROUP BY device_type
ORDER BY total_sessions DESC;
"""

df_q3 = pd.read_sql(query_q3, engine)

print("--- Q3: Device Analytics ---")
print(df_q3)

--- Q3: Device Analytics ---
  device_type  total_sessions  total_watch_min  avg_watch_min  completion_rate
0      Mobile           50172        1504355.0          29.98            60.24
1          TV           27981         840595.0          30.04            59.98
2      Laptop           15105         453434.0          30.02            60.51
3      Tablet            7091         210733.0          29.72            59.79


#### Q3 — Device analytics
**Answer:**  
 device_type  total_sessions  total_watch_min  avg_watch_min  completion_rate
      Mobile           50172        1504355.0          29.98            60.24
          TV           27981         840595.0          30.04            59.98
      Laptop           15105         453434.0          30.02            60.51
      Tablet            7091         210733.0          29.72            59.79
*
Grouped sessions by device_type.
Excluded rows where user_id IS NULL to ensure data accuracy.
Calculated COUNT, SUM, and AVG for sessions and minutes.
Computed completion rate using AVG(completed) * 100.
Rounded the final percentages to 2 decimal places.

In [ ]:
query_q4 = """
SELECT
    stars,
    COUNT(rating_id) AS rating_count,
    ROUND(COUNT(rating_id) * 100.0 / (SELECT COUNT(*) FROM ratings), 2) AS rating_pct
FROM ratings
GROUP BY stars
ORDER BY stars DESC;
"""

df_q4 = pd.read_sql(query_q4, engine)

print("--- Q4: Rating Distribution ---")
print(df_q4)

# Calculate the combined percentage for 4 and 5 stars
high_rating_pct = df_q4[df_q4['stars'].isin([4, 5])]['rating_pct'].sum()

print(f"\nPercentage of 4 or 5 stars: {high_rating_pct:.2f}%")

--- Q4: Rating Distribution ---
   stars  rating_count  rating_pct
0      5          1786       35.72
1      4          1781       35.62
2      3           847       16.94
3      2           352        7.04
4      1           234        4.68

Percentage of 4 or 5 stars: 71.34%


#### Q4 — Rating distribution
**Answer:**
  stars  rating_count  rating_pct
      5          1786       35.72
      4          1781       35.62
      3           847       16.94
      2           352        7.04
      1           234        4.68

**[Total % of 4 or 5 stars: 71.34%]**

Grouped the ratings by star value.
Calculated the count for each star.
Used a subquery to get the total rating count for percentage calculation.
Rounded percentages to 2 decimal places.
Summed the percentages of 4 and 5 stars for the final metric.

In [ ]:
query_q5 = """
SELECT
    CASE WHEN is_original = 1 THEN 'Originals' ELSE 'Acquired'
    END AS content_type,
    COUNT(show_id) AS no_shows,
    ROUND(AVG(imdb_rating), 2) AS avg_imdb_rating,
    ROUND(AVG(release_year), 0) AS avg_release_year
FROM shows
GROUP BY is_original, content_type
ORDER BY avg_imdb_rating DESC;
"""

df_q5 = pd.read_sql(query_q5, engine)

print("--- Q5: Originals vs Acquired ---")
print(df_q5)

# Calculate difference for the write-up
orig_rating = df_q5[df_q5['content_type'] == 'Originals']['avg_imdb_rating'].values[0]
acq_rating = df_q5[df_q5['content_type'] == 'Acquired']['avg_imdb_rating'].values[0]
diff = abs(orig_rating - acq_rating)

winner = 'Originals' if orig_rating > acq_rating else 'Acquired'
print(f"\n{winner} perform better by {diff:.2f} points.")

--- Q5: Originals vs Acquired ---
  content_type  no_shows  avg_imdb_rating  avg_release_year
0    Originals        30             7.92            2020.0
1     Acquired        70             6.63            2021.0

Originals perform better by 1.29 points.


#### Q5 — Originals vs acquired
**Answer:**
  content_type  no_shows  avg_imdb_rating  avg_release_year
0    Originals        30             7.92            2020.0
1     Acquired        70             6.63            2021.0

**Conclusion:**
Originals perform better on ratings.
The average rating is higher by 1.29 points.

**Logic:**
Grouped shows by original status.
Calculated total count and averages.
Rounded ratings to 2 decimals.
Compared the average ratings.
Identified the better performing category.

In [ ]:
query_q6 = """
SELECT
    user_id,
    show_id,
    session_date,
    COUNT(session_id) AS sessions_count
FROM watch_sessions
WHERE session_date BETWEEN '2024-04-01' AND '2024-06-30'
  AND user_id IS NOT NULL
GROUP BY user_id, show_id, session_date
HAVING COUNT(session_id) >= 5;
"""

df_q6 = pd.read_sql(query_q6, engine)

# Calculate the required metrics dynamically
total_binge_days = len(df_q6)
user_binge_counts = df_q6['user_id'].value_counts()
top_user = user_binge_counts.idxmax()
top_user_count = user_binge_counts.max()

print("--- Q6: Binge Day Detection ---")
print(f"Total binge days in Q2: {total_binge_days}")
print(f"User with MOST binge days: {top_user}")
print(f"Binge days for this user: {top_user_count}")

--- Q6: Binge Day Detection ---
Total binge days in Q2: 414
User with MOST binge days: U02956
Binge days for this user: 8


#### Q6 — Binge day detection
**Total binge days: 414** <br>
**Top user: U02956** <br>
**Top user's binge days: 8**

Filtered dates for Q2 2024.
Grouped by user, show, and date.
Kept rows with 5 or more sessions using HAVING.
Calculated total days and top user.

In [ ]:
query_q7 = """
SELECT
    COUNT(DISTINCT u.user_id) AS total_q1_signups,
    COUNT(DISTINCT CASE WHEN ws.session_id IS NULL THEN u.user_id END) AS never_watched
FROM users u
LEFT JOIN watch_sessions ws ON u.user_id = ws.user_id
WHERE MONTH(u.signup_date) <= 3
  AND YEAR(u.signup_date) = 2024;
"""

df_q7 = pd.read_sql(query_q7, engine)

print("--- Q7: Q1 Signups Who Never Watched ---")
print(df_q7)

--- Q7: Q1 Signups Who Never Watched ---
   total_q1_signups  never_watched
0              1250            226


#### Q7 — Q1 signups who never watched
**Answer: 226 users**<br>
**Total Q1 signups: 1250**

Filtered users for Q1 2024 signups.
Avoided the NOT IN trap due to NULL user_ids.
Used LEFT JOIN with watch_sessions.
Counted users where session_id IS NULL.
Used DISTINCT to prevent duplicate counts.

In [ ]:
query_q8 = """
WITH CurrentSubs AS (
    SELECT user_id, plan,
           ROW_NUMBER() OVER(PARTITION BY user_id ORDER BY start_date DESC) as rn
    FROM subscriptions
    WHERE status = 'active'
      AND start_date <= '2024-06-30'
      AND (end_date IS NULL OR end_date > '2024-06-30')
)
SELECT COUNT(cs.user_id) AS overpaying_users
FROM CurrentSubs cs
WHERE cs.rn = 1
  AND cs.plan IN ('Premium', 'Family')
  AND NOT EXISTS (
      SELECT 1
      FROM watch_sessions ws
      JOIN shows s ON ws.show_id = s.show_id
      WHERE ws.user_id = cs.user_id
        AND s.min_plan IN ('Premium', 'Family')
  );
"""

df_q8 = pd.read_sql(query_q8, engine)

print("--- Q8: Over-paying Users ---")
print(df_q8)

--- Q8: Over-paying Users ---
   overpaying_users
0               204


#### Q8 — The over-paying Premium/Family users
**Answer:204 users**

Used a CTE to find each user's current active plan as of June 30.
Filtered for users currently on Premium or Family plans.
Used NOT EXISTS to check their watch history.
Excluded any user who watched shows requiring Premium or Family tiers.
Counted the remaining over-paying users.

In [ ]:
query_q9 = """
WITH UserFirstSub AS (
    SELECT user_id, plan, start_date,
           ROW_NUMBER() OVER(PARTITION BY user_id ORDER BY start_date ASC) as rn
    FROM subscriptions
),
JanBasicCohort AS (
    SELECT u.user_id, u.signup_date
    FROM users u
    JOIN UserFirstSub ufs ON u.user_id = ufs.user_id
    WHERE YEAR(u.signup_date) = 2024 AND MONTH(u.signup_date) = 1
      AND ufs.rn = 1 AND ufs.plan = 'Basic'
),
FirstUpgrade AS (
    SELECT s.user_id, MIN(s.start_date) AS first_upgrade_date
    FROM subscriptions s
    JOIN JanBasicCohort jbc ON s.user_id = jbc.user_id
    WHERE s.plan IN ('Premium', 'Family')
    GROUP BY s.user_id
),
CurrentlyActive AS (
    SELECT DISTINCT user_id
    FROM subscriptions
    WHERE status = 'active'
      AND start_date <= '2024-06-30'
      AND (end_date IS NULL OR end_date > '2024-06-30')
)
SELECT
    COUNT(jbc.user_id) AS upgraded_users_count,
    ROUND(AVG(DATEDIFF(fu.first_upgrade_date, jbc.signup_date)), 2) AS avg_days_to_upgrade
FROM JanBasicCohort jbc
JOIN FirstUpgrade fu ON jbc.user_id = fu.user_id
JOIN CurrentlyActive ca ON jbc.user_id = ca.user_id;
"""

df_q9 = pd.read_sql(query_q9, engine)

print("--- Q9: Upgrade Success Cohort ---")
print(df_q9)

--- Q9: Upgrade Success Cohort ---
   upgraded_users_count  avg_days_to_upgrade
0                    55                64.96


#### Q9 — Upgrade success group
**Answer:55 users**<br>
**Average days to first upgrade:64.96 days**

Filtered January signups.
Used ROW_NUMBER() to confirm their very first plan was Basic.
Found the minimum start_date for Premium/Family to identify the first upgrade.
Checked active status as of June 30.
Calculated average DATEDIFF between signup and first upgrade.

In [ ]:
query_q10 = """
WITH ComebackEvents AS (
    SELECT DISTINCT
        ws1.user_id,
        ws1.show_id,
        ws1.session_date AS incomplete_date
    FROM watch_sessions ws1
    JOIN watch_sessions ws2
      ON ws1.user_id = ws2.user_id
     AND ws1.show_id = ws2.show_id
    WHERE ws1.completed = 0
      AND ws2.session_date BETWEEN DATE_ADD(ws1.session_date, INTERVAL 1 DAY)
                               AND DATE_ADD(ws1.session_date, INTERVAL 7 DAY)
      AND ws1.user_id IS NOT NULL
)
SELECT
    ce.show_id,
    s.title,
    COUNT(*) AS comeback_count
FROM ComebackEvents ce
JOIN shows s ON ce.show_id = s.show_id
GROUP BY ce.show_id, s.title
ORDER BY comeback_count DESC;
"""

df_q10 = pd.read_sql(query_q10, engine)

# Calculate total events and find the top show dynamically
total_events = df_q10['comeback_count'].sum()
top_show_id = df_q10.iloc[0]['show_id']
top_show_title = df_q10.iloc[0]['title']

print("--- Q10: Cliffhanger Comebacks ---")
print(f"Total Cliffhanger Events: {total_events}")
print(f"Top Show ID: {top_show_id}")
print(f"Top Show Title: {top_show_title}")

--- Q10: Cliffhanger Comebacks ---
Total Cliffhanger Events: 4345
Top Show ID: S088
Top Show Title: Rayalaseema Raga


#### Q10 — Cliffhanger comebacks
**Total comeback events: 4345**<br>
**Top show ID:S088**<br>
**Top show title: Rayalaseema Raga**

**Logic:**
Used a self-join on watch_sessions.
Matched the same user and show across both tables.
Filtered for first session completed = 0.
Used DATE_ADD to check if the second session was 1 to 7 days later.
Used DISTINCT to count each unique event only once.
Joined with the shows table to get the final title.

In [ ]:
query_q11 = """
WITH UserWeeks AS (
    SELECT DISTINCT
        user_id,
        WEEKOFYEAR(session_date) AS week_num
    FROM watch_sessions
    WHERE user_id IS NOT NULL
),
StreakGroups AS (
    SELECT
        user_id,
        week_num - ROW_NUMBER() OVER(PARTITION BY user_id ORDER BY week_num) AS grp
    FROM UserWeeks
),
UserStreaks AS (
    SELECT
        user_id,
        COUNT(*) AS streak_length
    FROM StreakGroups
    GROUP BY user_id, grp
)
SELECT
    user_id,
    MAX(streak_length) AS max_streak
FROM UserStreaks
GROUP BY user_id
HAVING MAX(streak_length) >= 4
ORDER BY max_streak DESC;
"""

df_q11 = pd.read_sql(query_q11, engine)
total_4_plus_users = len(df_q11)
longest_streak = df_q11['max_streak'].max()
top_streak_user = df_q11.iloc[0]['user_id']

print("--- Q11: Consecutive-week Engagement ---")
print(f"Users with 4+ weeks streak: {total_4_plus_users}")
print(f"Longest streak in weeks: {longest_streak}")
print(f"User with longest streak: {top_streak_user}")

--- Q11: Consecutive-week Engagement ---
Users with 4+ weeks streak: 1675
Longest streak in weeks: 26
User with longest streak: U00213


#### Q11 — Consecutive-week engagement
**Users with 4+ consecutive weeks: 1675**<br>
**Longest streak: 26 weeks**<br>
**User with longest streak:U00213**

Solved using the gaps-and-islands pattern.
Extracted distinct ISO week numbers for each user.
Calculated the difference between the week number and a sequential ROW_NUMBER.
Grouped by this difference to identify consecutive streaks.
Filtered for streaks >= 4 weeks.

In [ ]:
query_q12 = """
WITH MayJuneMins AS (
    SELECT
        ws.user_id,
        u.name,
        SUM(CASE WHEN MONTH(ws.session_date) = 5 THEN ws.watch_minutes ELSE 0 END) AS may_mins,
        SUM(CASE WHEN MONTH(ws.session_date) = 6 THEN ws.watch_minutes ELSE 0 END) AS june_mins
    FROM watch_sessions ws
    JOIN users u ON ws.user_id = u.user_id
    WHERE ws.session_date >= '2024-05-01'
      AND ws.session_date <= '2024-06-30'
      AND ws.user_id IS NOT NULL
    GROUP BY ws.user_id, u.name
)
SELECT
    user_id,
    name,
    may_mins,
    june_mins,
    ROUND(((may_mins - june_mins) / may_mins) * 100, 2) AS drop_percentage
FROM MayJuneMins
WHERE may_mins > 0
  AND ((may_mins - june_mins) / may_mins) >= 0.50
ORDER BY drop_percentage DESC;
"""

df_q12 = pd.read_sql(query_q12, engine)

print("--- Q12: Churn Signal Detection ---")
print(df_q12.head())
print(f"\nTOTAL Churn Signal Users: {len(df_q12)}")

--- Q12: Churn Signal Detection ---
  user_id               name  may_mins  june_mins  drop_percentage
0  U01931        Rohit Dutta     153.0        0.0            100.0
1  U01724      Yuvraj Prasad     120.0        0.0            100.0
2  U02220  Rohan Subramanian      68.0        0.0            100.0
3  U00446        Dhruv Singh     277.0        0.0            100.0
4  U00883       Atharv Verma     547.0        0.0            100.0

TOTAL Churn Signal Users: 521


#### Q12 — Churn signal detection
**TOTAL Churn Signal Users: 521**
Optional:  
 user_id               name  may_mins  june_mins  drop_percentage <br>
0  U01931        Rohit Dutta     153.0        0.0            100.0<br>
1  U01724      Yuvraj Prasad     120.0        0.0            100.0<br>
2  U02220  Rohan Subramanian      68.0        0.0            100.0<br>
3  U00446        Dhruv Singh     277.0        0.0            100.0<br>
4  U00883       Atharv Verma     547.0        0.0            100.0<br>

Created a CTE to calculate watch minutes for May and June using conditional aggregation.
This avoids the LAG() trap and gracefully handles users with 0 minutes in June.
Used an outer query to calculate the drop percentage dynamically.
Filtered for users who had > 0 minutes in May and a drop >= 50%.
Rounded the final drop percentage to 2 decimal places.